[feat] Add wait-ready command to check Multipass daemon readiness

rluna319 opened the pull request:
**Feature: Add `multipass wait-ready` command**

**Description:**
This PR introduces a new CLI command, `multipass wait-ready`. This command allows users and scripts to pause execution until the Multipass daemon is fully initialized and ready to accept requests. This is particularly beneficial for automation workflows that need to ensure the daemon is operational before proceeding with other Multipass operations.

The `wait-ready` command works by:
1.  Initially polling the daemon's gRPC socket until it becomes available.
2.  Once the socket is available, it sends a `WaitReady` request to the daemon.
3.  The daemon, upon receiving this request, verifies its readiness by attempting to update image manifests (which implicitly checks connectivity to image servers).
4.  The command supports a `--timeout` option (in seconds). If the daemon isn't ready within the specified timeout, the command exits with an error. If no timeout is given, it waits indefinitely.
5.  An animated spinner and context-relevant message provide user feedback during the waiting period.

**Motivation:**
- Improves the reliability of automated scripts and CI/CD pipelines that interact with Multipass by providing a deterministic way to wait for daemon initialization.
- Addresses scenarios where subsequent Multipass commands might fail if the daemon is still starting up.

**Related Issues:**
Fixes #452
Fixes #3723

**Key Implementation Changes:**
- **Client-side (`wait_ready.cpp`):** New command logic, including retry for socket connection, timeout handling (reusing `mp::cmd::parse_timeout`), and spinner integration.
- **Protocol (`multipass.proto`):** Added `WaitReadyRequest` / `WaitReadyReply` messages and `WaitReady` RPC method.
- **Daemon-side (`daemon_rpc.cpp`, `daemon.cpp`):**
    - New gRPC service handler in `daemon_rpc.cpp`.
    - Core readiness check in `daemon.cpp` leverages the existing `wait_update_manifests_all_and_optionally_applied_force()` function.

**How to Test:**
1.  Build this branch and `$ cd build`.
2.  Ensure the Multipass daemon is stopped.
3.  To verify `--timeout`, simply run the `wait-ready` command with the `--timeout` option:
```Bash
$ ./bin/multipass wait-ready --timeout 5
```
> Be sure not to start up the daemon first or the command might exit before the timeout.
- Expected: Info message "Waiting for Multipass daemon to be ready...", spinner is shown, command times out after ~5 seconds with an appropriate error message "Timed out waiting for Multipass daemon to be ready."
4. (Optional) To replicate the issue that happens when the client tries to use the daemon before its ready (without `wait-ready`), start up the daemon and immediately use `multipass launch`:
```Bash
$ sudo -v
$ (sudo ./bin/multipassd &) && ./bin/multipass launch
```
- Expected: Error message "launch failed: cannot connect to the multipass socket", "launch failed: Remote "" is unknown or unreachable.", or something like issue #3723
5.  Stop the Multipass daemon
```Bash
$ sudo pkill multipassd
```
6. To verify `wait-ready` command solves the issue, fit the `wait-ready` command between daemon startup and `multipass launch`:
```Bash
$ sudo -v
$ (sudo ./bin/multipassd &) && ./bin/multipass wait-ready && ./bin/multipass launch
```
> Optionally, add `--timeout` if desired.
- Expected: Info message "Waiting for Multipass daemon to be ready...", spinner is shown, command exits successfully (ReturnCode::Ok) once the daemon is ready (i.e., can connect to image servers) and the client begins retrieving the default _release_ image for the `multipass launch` process. _You may Ctrl+C here to terminate the launch process._

**Points for Reviewers:**
- Confirmation that using `wait_update_manifests_all_and_optionally_applied_force(true)` is an appropriate and robust proxy for overall daemon readiness.
- Feedback on the client-side error messages and user experience (spinner, timeout message) is appreciated. The current spinner implementation addresses @ricab's suggestion in #452.
- **Unit Tests:** Per @ricab 's comments in issue #452, unit tests have not yet been implemented for this initial PR to facilitate earlier review. I am prepared to add them based on feedback.

ricab commented:
Hey @rluna319, thank you for this! We'll review when we get a chance.

levkropp reviewed (dismissed):
Yep, tested and working with the examples provided. Great additions!

Needs a pass through `clang-format-16` to satisfy the linter I believe

xmkg reviewed (changes requested):
Great work, @rluna319! I appreciate the work you've put into this. I was on the verge of implementing a similar feature for the upcoming CLI tests I'm working on, and you can't imagine how happy I was to see this PR :)

Overall, the code looks good! I built and tested your changes locally, and everything works as expected. I have a small request in line, and also, having some unit tests would be nice.

rluna319 commented on src/daemon/daemon.cpp:
```diff
@@ -3127,6 +3128,41 @@ catch (const std::exception& e)
+    // If the force_manifest_network_download is true, it will download the manifests even if already cached.
+    try 
+    {
+        wait_update_manifests_all_and_optionally_applied_force(/*force_manifest_network_download=*/true);
```
@xmkg Thank you for the review!

> I'm skeptical about forcing a manifest download here. The daemon's constructor already dispatches a manifest update by constructing the `update_manifests_all_task` object, so wouldn't `force_manifest_network_download=false` be enough to ensure daemon readiness?
> 
> ~I do believe we should have an option to force it, though. In that case, we can mimic what `find` is already doing, i.e., `--force-update`.~ Let me know if I'm missing something.

This was a concern of mine when deciding to reuse this function out of its intended use case. The goal of forcing a manifest download here for the `wait-ready` is to ensure that the daemon can reach the image servers (solving #3723) not entirely to force the manifests to be downloaded/updated.

The `update_manifests_all_task` is started in the background but to my understanding its an asynchronous task dispatched to Qt and I don't think there is any way to check that it is completed before the client issues a command. 

If the cache is valid on daemon startup then `force_manifest_network_download=false` should work. But in cases where the daemon needs to fetch the manifests and connect to the image server, i don't think that would be enough.

I guess it would really depend on our exact definition of "readiness." Checking that the daemon is simply reachable is one way to approach that. But to cover all bases and prevent errors like #3723, I think the wait-ready should include waiting for the daemon to be able to connect to the image servers as well. 

****

In the event that the user is working offline and has the image cached or is using custom images, always checking connectivity to image servers is NOT desirable. I had not accounted for this until right now. So this is a fix that should be implemented for sure. 

I haven't explored a more custom method of checking image server connectivity. There might be a cleaner way of doing this. 

And just as well, I suppose a quick and dirty workaround would be to just have `force_manifest_network_download=false` and after `wait-ready` use `find --force-update` (if the user requires an image from the image servers) before using `launch`.

ricab replied:
Actually, after thinking a little more about this, I think I would rather simplify and just have `wait-ready` complete successfully only after the manifests have been successfully updated. Yes, people can use the daemon offline with cached images, but nothing in here prevents that. They would just not be able to use the wait-ready functionality because, in a sense, the daemon can't be fully ready until it retrieves image info.

> I guess it would really depend on our exact definition of "readiness." 

Exactly, and I think the definition of ready that makes most sense is "fully ready for normal operation", which requires online connectivity and updated manifests.

> Perhaps it might also be a good idea to call `wait_update_manifests_all_and_optionally_applied_force(false)` in functions that depend on the manifests, such as `launch`.

It might be a good idea, as long as we updated the spinner message while doing so, because it can take a long time if there are connectivity issues (exponential back-off). And the launch command would then just hang. As it is, the launch just fails in such circumstances. That would be for a different PR though.

ricab replied:
If the command was run after 15min (period for the manifest update), it would again hang for a little bit. It's no big deal but, to be very precise, you could add an accessor to `AsyncPeriodicDownloadTask` reporting whether the task had ever completed. Then replace the call to `mp::Daemon::wait_update_manifests_all_and_optionally_applied_force` with something like `update_manifests_all_task.ever_completed() || update_manifests_all_task.wait_ongoing_task_finish()`. Or perhaps even better, just add a method to `wait_for_first_completion()`. Names up for improvement of course.

ricab commented:
@rluna319, could you also rebase on the latest main, to see if we can get some CI to run on this? Appreciate it.

rluna319 commented:
@ricab I had not included your requested changes on my latest commit. I had not seen them before I made the push, my fault. I did rebase on the latest main however.

I will work on your requested changes ASAP.

Would you like me to amend your requested changes to my latest commit or have it in a separate commit?

ricab commented:
Hey @rluna319 no worries. We generally prefer additional commits, to make it easy to identify new changes. Thanks!

rluna319 commented on src/client/cli/cmd/wait_ready.cpp:
```diff
@@ -0,0 +1,131 @@
+{
+    mp::cmd::add_timeout(parser);
+
+    auto status = parser->commandParse(this);
```
Sure, would simply removing the portion concerning instances suffice? More or less leaving it consistent with its description in the other commands with the `--timeout` option.

"Maximum time, in seconds, to wait for the command to complete. Note that some background operations may continue beyond that."

Or we could tailor it exclusively for the wait-ready with something like:

"Maximum time, in seconds, to wait for the Multipass daemon to be ready. Note that some background operations may continue beyond that."

rluna319 replied:
Also, in the original help text. There is mention of a `default_timeout`. This is strictly regarding the instances, not concerning `--timeout` arg itself, correct? We can safely disregard this in our adapted help text?

ricab replied:
Right, I suggest one of two approaches:

- Easy: dedicated message here and rename `add_timeout` to `add_instance_timeout`
- Nice: split the instance-specific part of `add_timeout` into `add_instance_timeout` (calling the former and completing it); then call add_timeout in here

ricab reviewed (commented):
Getting real close @rluna319! I am going to let CI go through and I think the Lint workflow is going to request some formatting changes. Other than that, I have mostly minor requests. There's only one for real behavior, but I think we can discuss and split off as a separate task if we need it.

rluna319 commented on src/daemon/daemon.cpp:
```diff
@@ -3130,6 +3131,41 @@ catch (const std::exception& e)
+    catch(const std::exception& e)
+    {
+        logger.log(mpl::Level::error, "daemon", fmt::format("Failed to connect to image servers: {}", e.what()));
+        status_promise->set_value(grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, e.what(), ""));
```
>If it kept waiting, the daemon might be able to connect eventually. So, in an environment that took a while to connect, people would have run wait-ready in a loop, which kind of defeats the purpose, at least for the disconnected use case.

This was something I meant to ask about and address on the second commit. It slipped my mind.

Setting the return code in the inner try catch block  from `FAILED_PRECONDITION` to `UNAVAILABLE` (what I had previously but unintentional at the time) would have it retry instead. And then we also just change the log level to debug instead of error.

>OTOH, this PR has been going for a while and I don't want to ask for much more. It's certainly valuable even so and I am thinking we could probably improve this aspect at a later time, if there was demand for it.

Also sounds good to me. I don't mind doing a bit more either. I much appreciate the experience this contribution is providing me.

rluna319 replied:
@ricab @xmkg we could do something like this:

```
// We use wait_update_manifests_all_and_optionally_applied_force to check connectivity to image servers.
    while (true)
    {
        try
        {
            wait_update_manifests_all_and_optionally_applied_force(/*force_manifest_network_download=*/false);
            logger.log(mpl::Level::debug, "daemon", "Successfully connected to image servers.");
            status_promise->set_value(grpc::Status::OK);
            return; // Success, exit the function.
        }
        catch(const mp::DownloadException& e)
        {
            logger.log(mpl::Level::debug, "daemon", fmt::format("Failed to connect to image servers, retrying...: {}", e.what()));
            // The client is responsible for the timeout, so we just wait and retry.
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
    }
```

The outer catch block should catch any other exception. I believe the only exception we care about concerning the connection to the image servers is `DownloadException` which we would just retry on till we reach the `--timeout` if specified.

rluna319 replied:
I'll leave that for later if it's desired.

ricab commented:
Hey @rluna319, it looks like you [haven't signed our CLA](https://github.com/canonical/multipass/actions/runs/16452896886/job/46735070225?pr=4145) yet. Would you mind doing so, please? https://ubuntu.com/legal/contributors

Also, there are a few [linting problems](https://github.com/canonical/multipass/actions/runs/16452897071/job/46735084487?pr=4145). Editors often have options to strip trailing whitespace and keep a single newline at the end. You can check whitespace sanity on a diff with `git diff --check ...`. For example: `git diff --check main...wait-ready`.

`git clang-format` might take care of whitespaces too, if you set it up. That might be a good idea anyway because it's going to run once the whitespace check succeeds.

rluna319 commented:
@ricab I signed the CLA, went through linter errors (and other CI build/test errors) and fixed them to the best of my knowledge, ran the code through `git clang-format` and `git diff --check main`. I've also rebased off the latest main.

rluna319 commented:
I removed some commented out includes that shouldn't have been in that commit.

rluna319 commented on 3rd-party/jsoncpp:
https://github.com/canonical/multipass/blob/db41306c7f7570baa3ae9652f2f6a726fed90fd2/.gitmodules#L17

I think main currently tracks jsoncpp as a submodule. Am I rebasing wrong?

rluna319 commented on src/client/cli/cmd/wait_ready.cpp:
```diff
@@ -0,0 +1,135 @@
+
+    auto on_failure = [this, &spinner, &timer](grpc::Status& status) {
+        if (status.error_code() == grpc::StatusCode::NOT_FOUND &&
+            (status.error_message() == "cannot connect to the multipass socket" ||
+             status.error_message() == "cannot connect to the image servers"))
```
It goes back to what I was trying to say about using `UNAVAILABLE` before I decided that `NOT_FOUND` would be more suitable. If we were to implement the `--interval` option, the specified interval would be used for both waiting for the daemon socket to become available and waiting for the daemon to connect to the image servers.

If we set the status to `UNAVAILABLE`, the dispatcher will change this to `NOT_FOUND` with the "cannot connect to the multipass socket" string. Using `NOT_FOUND` lets us at least set the error string to a more accurate message about image server connectivity. After that the goal was to be able to have `--interval` only affect the polling for the image servers, not socket availability. So the only way to differentiate in wait_ready cpp would be to check the error string. 

 >I don't think clients should depend on message strings for behavior (that's what the error code is for).
 
 I agree. This is probably not a good approach thinking back on it. Especially now that we aren't implementing the `--interval` option yet.
 
 >Is there any other case that would fall under NOT_FOUND
 
 Regarding wait_ready no not that i'm aware of. 
 
 We should be able to safely get rid of also checking the error messages.

rluna319 commented:
I see that I still missed some linter errors. I thought I ran everything through `clang-format` correctly. I'll address those.

rluna319 commented:
@ricab Ive addressed the issues. It says some of the workflows are awaiting approval to be run.

xmkg commented:
> @ricab Ive addressed the issues. It says some of the workflows are awaiting approval to be run.

Approved the workflow.

xmkg commented:
> Thank you @xmkg.
> 
> I think that the remaining workflows that fail are due to some internal CI issues. Something about a missing token. Is this something I need to fix or can fix on my end?

Yeah, unfortunately, our CI is not happy with PRs coming from external forks. We're working on making this process smoother, but the necessary machinery is not yet in place. Would you mind if I create a PR for your PR, just for the sake of running the pipeline? We can track the progress and make changes here.

rluna319 commented:
Oh ok. Sure thing.

xmkg commented:
@rluna319, we've landed a bunch of CI-related fixes to `main`. Could you rebase this PR?

xmkg commented:
> @xmkg I have rebased the PR. Workflows awaiting approval.

Great, thanks. Synced #4304.

rluna319 commented:
@xmkg Great thank you. I see all the tests have passed there. Should I mind the coverage report? I've looked at it already and I can cover the line in `wait_ready.cpp` that it says isn't covered. 

However, i'm having trouble covering the other one in `daemon.cpp` where we catch any other std::exception. Throwing a runtime error in update_manifests just causes the test to hang. I'm guessing its the way that exception is or isn't handled in the asynchronous task. I'm not sure what other exception would make it to that catch block. Any other exceptions that ive traced within the codepath are already handled locally. Perhaps i'm just missing something i'm not aware of.

xmkg reviewed (approved):
Looks good to me, thanks @rluna319!

ricab reviewed (approved):
I just did a final pass and it's looking great. Thanks again!

ricab commented:
Our pleasure @rluna319! This was a great contribution, very well pondered. Feel free to send more in!
